### ISSUE TITLE
Field Extensions and DuplicateNamesError

### ISSUE THREAD

**Issue by cocoahero** [rmosolgo/graphql-ruby#3743]
We've noticed that a somewhat common scenario in our codebase is causing problems with the new [dynamic schema implementation](https://github.com/rmosolgo/graphql-ruby/pull/3651) when it comes to field extensions. Below is a somewhat contrived example:

```ruby
class SearchableExtension < GraphQL::Schema::FieldExtension
  def apply
    field.argument(:query, String, required: false)
  end
end

class QueryRoot < GraphQL::Schema::Object
  field :search, [String], null: false, extensions: [SearchableExtension] do |field|
    field.argument(:query, String, required: true, description: "Override description")
  end
end
```

This schema results in a `DuplicateNamesError` because the field is being declared twice. This is slightly different behavior from when a type is subclassed where the field is overwritten (which I think is the desired outcome here).

**Comment by cocoahero** [comment]
Wondering if a simple `override_extension` flag would suffice here, to be either used in warden, or our own `visible?` implementation.

**Comment by cocoahero** [comment]
A somewhat related scenario which also has this problem is connection fields:
```ruby
class ArticleType < GraphQL::Schema::Object
  field :title, String, null: false
end

class QueryRoot < GraphQL::Schema::Object
  field :articles, ArticleType.connection_type, null: false do |field|
    field.argument(:first, Integer, required: false, description: "Returns the first _N_ articles.")
  end

  def articles
    Article.all
  end
end
```
```
GraphQL::Schema::DuplicateNamesError (Found two visible definitions for `QueryRoot.articles.first`: 
#<GraphQL::Schema::Argument QueryRoot.articles.first: Int @description="Returns the first _n_ elements from the list.">, 
#<GraphQL::Schema::Argument QueryRoot.articles.first: Int @description="Returns the first _N_ articles.">)
```

**Comment by rmosolgo** [comment]
Thanks for reporting this issue ... I didn't think of this possible incompatibility 😖 . I don't know how GraphQL-Ruby could just "figure it out" ... When the definitions are different, how can it tell if one should _replace_ the other or if they're conditional?

The _easiest_ fix I can think of would be to require `override: true` when a field or argument (or enum value?) should _replace_ the previous definition with that name. (It would fail if there were multiple definitions with that name.) 

Alternatively, it could work the other way: an argument or field would override by default, and it could require a certain option (`multidefinition: true`) to define a parallel definition for something. 

What do you think of those approaches? Would one (or more) work for you? Anything else you think we could try?

**Comment by cocoahero** [comment]
> The easiest fix I can think of would be to require override: true when a field or argument (or enum value?) should replace the previous definition with that name. (It would fail if there were multiple definitions with that name.)

That was an idea I had as well - although it kind of feels like CSS's `!important` and chaos that comes with it 😅.

I'm not sure of a better solution, other than treating extensions as special and tracking the modifications they make in `apply_to` and having some logic similar to subclassing for automatic overrides.

**Comment by rmosolgo** [comment]
> treating extensions as special

That might work fine, especially since both cases identified about boil down to that 🤔 . 

Maybe adding a method which creates an argument which defers to an overridden one: 

```ruby
def apply
  field.default_argument(...) 
end 
```

Or, applying extensions _after_ definition blocks: 

```ruby 
def apply
  if !field.all_argument_definitions.any? { |arg| arg.keyword == :first }
    field.argument(:first, ...)
  end 
  # and so on 
end 
```

Or, letting extensions define default arguments:

```ruby 
class ConnectionExtension < FieldExtension
  default_argument :first, ... 
  default_argument :after, ... 
end 
```

Which does one of those ☝️ under the hood. 🤔 Your thoughts?

**Comment by rmosolgo** [comment]
I think this was addressed in #3751 -- if I missed something, or if you give it a try and it doesn't work, let me know!
